src/ratbag_emu/protocol/base.py

```diff
@@ -201,31 +201,37 @@ def _send_packets(self, packets, total):
 
 
 class BaseDevice(object):
-    def __init__(self, rdesc=None, info=None, name='Generic Device',
-                 shortname='generic'):
-        self.info = info
-        self.name = name
-        self.shortname = shortname
+    def __init__(self, obj):
+        self.info = tuple(obj['bus_info'])
+        self.name = obj['name']
+        self.shortname = obj['shortname']
         self.endpoints = {}
 
-        self.endpoints[0] = SimulatedDevice(self, rdesc, info, name, shortname)
+        for i, rdesc in obj['report_descriptors'].items():
```

Comment:
The report descriptor should not be embedded in the json object but stored in the protocol class or in the `BaseDevice` class (as an attribute you will be overwriting).

Reply:
The report descriptor is part of the device definition, I don't think it should be stored separately. If it helps you could think as an initial value, just like the LEDs or DPI values. Or do you think we should have that separately too? And just like before, the use case where we would be modifying the report descriptor is not really included in the main motivation behind ratbag-emu.

Reply:
No, the report descriptor should not be included in the json sent over the wire.

> If it helps you could think as an initial value, just like the LEDs or DPI values. 

No, that's 2 rather different concepts. One is written in the ROM of the mouse, the others are values that the user can set.

> Or do you think we should have that separately too?

No, and for the simple reason that you can get exactly what we want in https://github.com/libratbag/libratbag/commit/62ac0772f572ba674ed88373169c0587d5089934

> And just like before, the use case where we would be modifying the report descriptor is not really included in the main motivation behind ratbag-emu.

I mentioned *once* about dynamically generating the report descriptor, and I then said that you could bypass this step for now.
I am not talking about generating a report descriptor, I am talking about where to store the information. And the information should be in the server, not in the client.